eventcrypto.go

```diff
@@ -0,0 +1,161 @@
+package gomatrixserverlib
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"golang.org/x/crypto/ed25519"
+)
+
+// AddContentHashesToEvent sets the "hashes" key of the event with a SHA-256 hash of the unredacted event content.
+// This hash is used to detect whether the unredacted content of the event is valid.
+func AddContentHashesToEvent(eventJSON []byte) ([]byte, error) {
+	var event map[string]rawJSON
+
+	if err := json.Unmarshal(eventJSON, &event); err != nil {
+		return nil, err
+	}
+
+	unsignedJSON := event["unsigned"]
+
+	delete(event, "signatures")
```

Comment:
You never add this back in to `event`. Why? You add `unsigned` back in, and work out `hashes` so this seems like a bug or a subtlety which needs explaining.